### PR TITLE
[12.0][MIG] password_security

### PR DESCRIPTION
**Pull request by shepilov-vladislav**
Continue https://github.com/OCA/server-auth/pull/66

### CODE REVIEW

**Comment by pedrobaeza** [comment]
@shepilov-vladislav you have to preserve commit history as told before, and Travis is red in this PR. Please amend both things.

**Comment by pedrobaeza** [comment]
Full commit history is brought this way: https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-12.0#technical-method-to-migrate-a-module-from-110-to-120-branch

**Comment by shepilov-vladislav** [comment]
@pedrobaeza I have a problem with one test. How I can setup environment like on travis? With my environment all works fine.

File: password_security/__manifest__.py [hunk 1]
```diff
@@ -1,29 +1,26 @@
...
-    ],
-    "website": "https://laslabs.com",
+    "version": "12.0.1.1.0",
+    "author": "LasLabs, "
+    "Kaushal Prajapati, "
+    "Tecnativa, "
+    "Modoolar, "
+    "Odoo Community Association (OCA)",
+    "category": "Base",
+    "depends": ["auth_signup", "auth_password_policy_signup"],
```
Comment (yajo) [inline password_security/__manifest__.py]: Please avoid all of these cosmetic changes. I see people changing from double to single quotes and viceversa all the time. Just don't touch if you don't need to please.

File: password_security/__manifest__.py [hunk 2]
```diff
@@ -1,29 +1,26 @@
...
-        'views/res_company_view.xml',
-        'security/ir.model.access.csv',
-        'security/res_users_pass_history.xml',
-    ],
-    "demo": [
-        'demo/res_users.xml',
+        "views/password_security.xml",
+        "views/res_config_settings_views.xml",
+        "security/ir.model.access.csv",
+        "security/res_users_pass_history.xml",
```
Comment (yajo) [inline password_security/__manifest__.py]: Same here.

File: password_security/controllers/__init__.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+# pylint: disable=missing-docstring
```
Comment (yajo) [inline password_security/controllers/__init__.py]: This shouldn't be needed. Remove this line. If it fails, then it should be fixed in MQT.

File: password_security/controllers/main.py [hunk 4]
```diff
@@ -1,39 +1,41 @@
+# pylint: disable=missing-docstring
```
Comment (yajo) [inline password_security/controllers/main.py]: Same here.

File: password_security/controllers/main.py [hunk 5]
```diff
@@ -1,39 +1,41 @@
...
-from odoo.http import request
 from odoo.addons.auth_signup.controllers.main import AuthSignupHome
-from odoo.addons.web.controllers.main import ensure_db, Session
+from odoo.addons.web.controllers.main import Session, ensure_db
+from odoo.http import request
 
 from ..exceptions import PassError
 
 
+# pylint: disable=too-few-public-methods,protected-access
```
Comment (yajo) [inline password_security/controllers/main.py]: Same here. What pylint config are you using? :thinking: 

See @OCA's configs in https://github.com/OCA/maintainer-quality-tools/tree/master/travis/cfg.

File: password_security/controllers/main.py [hunk 6]
```diff
@@ -1,39 +1,41 @@
...
 class PasswordSecuritySession(Session):
-
     @http.route()
     def change_password(self, fields):
-        new_password = operator.itemgetter('new_password')(
-            dict(list(map(operator.itemgetter('name', 'value'), fields)))
+        new_password = operator.itemgetter("new_password")(
+            dict(list(map(operator.itemgetter("name", "value"), fields)))
```
Comment (yajo) [inline password_security/controllers/main.py]: Undo this change.

File: password_security/controllers/main.py [hunk 7]
```diff
@@ -1,39 +1,41 @@
...
         user_id._check_password(new_password)
         return super(PasswordSecuritySession, self).change_password(fields)
 
 
+# pylint: disable=protected-access
 class PasswordSecurityHome(AuthSignupHome):
-
     def do_signup(self, qcontext):
-        password = qcontext.get('password')
+        password = qcontext.get("password")
```
Comment (yajo) [inline password_security/controllers/main.py]: Here too.

File: password_security/controllers/main.py [hunk 8]
```diff
@@ -1,39 +1,41 @@
...
-            dict(list(map(operator.itemgetter('name', 'value'), fields)))
+        new_password = operator.itemgetter("new_password")(
+            dict(list(map(operator.itemgetter("name", "value"), fields)))
         )
         user_id = request.env.user
         user_id._check_password(new_password)
         return super(PasswordSecuritySession, self).change_password(fields)
 
 
+# pylint: disable=protected-access
```
Comment (yajo) [inline password_security/controllers/main.py]: Remove.

File: password_security/controllers/main.py [hunk 9]
```diff
@@ -1,39 +1,41 @@
...
 class PasswordSecurityHome(AuthSignupHome):
-
```
Comment (yajo) [inline password_security/controllers/main.py]: Restore.

File: password_security/controllers/main.py [hunk 10]
```diff
@@ -1,39 +1,41 @@
...
-        password = qcontext.get('password')
+        password = qcontext.get("password")
         user_id = request.env.user
         user_id._check_password(password)
         return super(PasswordSecurityHome, self).do_signup(qcontext)
 
     @http.route()
     def web_login(self, *args, **kw):
         ensure_db()
+
```
Comment (yajo) [inline password_security/controllers/main.py]: Remove.

File: password_security/controllers/main.py [hunk 11]
```diff
@@ -52,10 +54,10 @@ def web_auth_signup(self, *args, **kw):
             return super(PasswordSecurityHome, self).web_auth_signup(
                 *args, **kw
             )
-        except PassError as e:
+        except PassError as exc:
             qcontext = self.get_auth_signup_qcontext()
-            qcontext['error'] = e.message
-            return request.render('auth_signup.signup', qcontext)
+            qcontext["error"] = str(exc)
+            return request.render("auth_signup.signup", qcontext)
```
Comment (yajo) [inline password_security/controllers/main.py]: Undo uneeded diff.

File: password_security/controllers/main.py [hunk 12]
```diff
@@ -65,20 +67,18 @@ def web_auth_reset_password(self, *args, **kw):
...
         qcontext = self.get_auth_signup_qcontext()
         if (
-            request.httprequest.method == 'POST' and
-            qcontext.get('login') and
-            'error' not in qcontext and
-            'token' not in qcontext
+            request.httprequest.method == "POST"
+            and qcontext.get("login")
+            and "error" not in qcontext
+            and "token" not in qcontext
```
Comment (yajo) [inline password_security/controllers/main.py]: Same here.

File: password_security/controllers/main.py [hunk 13]
```diff
@@ -65,20 +67,18 @@ def web_auth_reset_password(self, *args, **kw):
...
         ):
-            login = qcontext.get('login')
+            login = qcontext.get("login")
```
Comment (yajo) [inline password_security/controllers/main.py]: Undo.

File: password_security/controllers/main.py [hunk 14]
```diff
@@ -65,20 +67,18 @@ def web_auth_reset_password(self, *args, **kw):
...
             user_ids = request.env.sudo().search(
-                [('login', '=', login)],
-                limit=1,
+                [("login", "=", login)], limit=1
```
Comment (yajo) [inline password_security/controllers/main.py]: Undo

File: password_security/controllers/main.py [hunk 15]
```diff
@@ -65,20 +67,18 @@ def web_auth_reset_password(self, *args, **kw):
...
             )
             if not user_ids:
                 user_ids = request.env.sudo().search(
-                    [('email', '=', login)],
-                    limit=1,
+                    [("email", "=", login)], limit=1
```
Comment (yajo) [inline password_security/controllers/main.py]: Undo

File: password_security/exceptions.py [hunk 16]
```diff
@@ -1,11 +1,14 @@
+# pylint: disable=missing-docstring
```
Comment (yajo) [inline password_security/exceptions.py]: Undo

File: password_security/exceptions.py [hunk 17]
```diff
@@ -1,11 +1,14 @@
...
 # Copyright 2015 LasLabs Inc.
 # License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
 
 from odoo.exceptions import Warning as UserError
 
 
+# pylint: disable=too-few-public-methods
```
Comment (yajo) [inline password_security/exceptions.py]: Undo

File: password_security/exceptions.py [hunk 18]
```diff
@@ -1,11 +1,14 @@
...
 class PassError(UserError):
     """ Example: When you try to create an insecure password."""
+
```
Comment (yajo) [inline password_security/exceptions.py]: Undo

File: password_security/models/__init__.py [hunk 19]
```diff
@@ -1,6 +1,10 @@
...
 
-from . import res_users
-from . import res_company
-from . import res_users_pass_history
+from . import (
+    res_company,
+    res_config_settings,
+    res_users,
+    res_users_pass_history,
+)
```
Comment (yajo) [inline password_security/models/__init__.py]: Undo

File: password_security/models/__init__.py [hunk 20]
```diff
@@ -1,6 +1,10 @@
+# pylint: disable=missing-docstring
```
Comment (yajo) [inline password_security/models/__init__.py]: Undo

File: password_security/models/res_company.py [hunk 21]
```diff
@@ -1,51 +1,43 @@
+# pylint: disable=missing-docstring
```
Comment (yajo) [inline password_security/models/res_company.py]: Undo

File: password_security/models/res_company.py [hunk 22]
```diff
@@ -1,51 +1,43 @@
...
 # License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
 
-from odoo import models, fields
+from odoo import fields, models
 
 
+# pylint: disable=too-few-public-methods
 class ResCompany(models.Model):
-    _inherit = 'res.company'
+    _inherit = "res.company"
```
Comment (yajo) [inline password_security/models/res_company.py]: Undo

**Review by yajo — CHANGES_REQUESTED** [review]
I'm reviewing the big diff introduced by this patch and it all seems to come from basically aesthetic styling.

We have a linter to do that job. Please do not "fix" any line that does not fail in a linter (there's a LINT job in travis that you can check).

I commented on 22 of them, but got tired after. Maybe you have a autostyler installed? Please undo all those "fixes" to review again.

Thanks!

**Comment by shepilov-vladislav** [comment]
> I'm reviewing the big diff introduced by this patch and it all seems to come from basically aesthetic styling.
> 
> We have a linter to do that job. Please do not "fix" any line that does not fail in a linter (there's a LINT job in travis that you can check).
> 
> I commented on 22 of them, but got tired after. Maybe you have a autostyler installed? Please undo all those "fixes" to review again.
> 
> Thanks!

@Yajo I corrected everything you asked for.

File: password_security/models/res_config_settings.py [hunk 23]
```diff
@@ -0,0 +1,37 @@
+# pylint: disable=missing-docstring
```
Comment (yajo) [inline password_security/models/res_config_settings.py]: :fire: Remove this line please.

File: password_security/models/res_config_settings.py [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+# Copyright 2018 Modoolar <info@modoolar.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+from odoo import fields, models
+
+
+# pylint: disable=too-few-public-methods
```
Comment (yajo) [inline password_security/models/res_config_settings.py]: :fire:

File: password_security/models/res_config_settings.py [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    password_expiration = fields.Integer(
+        related="company_id.password_expiration", readonly=False
+    )
+
```
Comment (yajo) [inline password_security/models/res_config_settings.py]: :fire: Remove this empty line please.

File: password_security/models/res_config_settings.py [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+    password_minimum = fields.Integer(
+        related="company_id.password_minimum", readonly=False
+    )
+
```
Comment (yajo) [inline password_security/models/res_config_settings.py]: :fire:

File: password_security/models/res_config_settings.py [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+    password_history = fields.Integer(
+        related="company_id.password_history", readonly=False
+    )
+
```
Comment (yajo) [inline password_security/models/res_config_settings.py]: :fire:

File: password_security/models/res_users.py [hunk 28]
```diff
@@ -13,7 +14,7 @@
 
 def delta_now(**kwargs):
     dt = datetime.now() + timedelta(**kwargs)
-    return fields.Datetime.to_string(dt)
+    return fields.Datetime.to_datetime(dt)
```
Comment (yajo) [inline password_security/models/res_users.py]: I think this is already a datetime, isn't it? Shouldn't this be then just `return dt`?

File: password_security/static/src/js/password_gauge.js [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+            var specialscore = this._calculate_password_score(
+                "[\\W_]", this._password_special, password);
+
+            return (
+                lengthscore * loverscore * upperscore *
+                numericscore * specialscore);
+        },
+    });
+
+    require('auth_password_policy').recommendations.policies = [
```
Comment (yajo) [inline password_security/static/src/js/password_gauge.js]: Require it only once at the top please.

**Review by yajo — CHANGES_REQUESTED** [review]
Some minor changes. Mostly code is OK. Thanks!

File: password_security/models/res_users.py (hunk 28 above)
Comment (shepilov-vladislav) [reply password_security/models/res_users.py]: It's my bad. Fixed.

**Review by yajo — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Travis is red

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Can you use new README by fragments system?

Also version should be 12.0.1.0.0.

And finally, please stay with only one migration commit.

File: password_security/models/res_users.py [hunk 30]
```diff
@@ -97,8 +123,12 @@ def _password_has_expired(self):
...
         if not self.password_write_date:
             return True
-        write_date = fields.Datetime.from_string(self.password_write_date)
-        today = fields.Datetime.from_string(fields.Datetime.now())
+
+        if not self.company_id.password_expiration:
+            return False
+
+        write_date = fields.Datetime.to_datetime(self.password_write_date)
+        today = fields.Datetime.to_datetime(fields.Datetime.now())
```
Comment (ernestotejeda) [inline password_security/models/res_users.py]: ```self.password_write_date``` and ```fields.Datetime.now()``` are already datetime objects

File: password_security/models/res_users.py [hunk 31]
```diff
@@ -120,7 +150,7 @@ def _validate_pass_reset(self):
             pass_min = rec_id.company_id.password_minimum
             if pass_min <= 0:
                 pass
-            write_date = fields.Datetime.from_string(
+            write_date = fields.Datetime.to_datetime(
```
Comment (ernestotejeda) [inline password_security/models/res_users.py]: the same here: ```rec_id.password_write_date``` is already a datetime object

**Review by ernestotejeda — CHANGES_REQUESTED** [review]
Some change requests

**Comment by shepilov-vladislav** [comment]
Guys, can you help me. I want to fix my one failed test. But I can not reproduce it. Could you give me instructions on how to raise the same environmental locally for tests(like in travis)? Maybe OCA has docker images for it?

**Comment by pedrobaeza** [comment]
Do you mean it's not reproducible in local?

**Comment by shepilov-vladislav** [comment]
> Do you mean it's not reproducible in local?

@pedrobaeza  Exactly. How I can reproduce environment like used on travis?

**Comment by pedrobaeza** [comment]
You can check the log for seeing how Travis is installing first "base" modules, and then installing the rest with --test-enable flag. You can produce the same using that 2 commands.

**Comment by shepilov-vladislav** [comment]
> You can check the log for seeing how Travis is installing first "base" modules, and then installing the rest with --test-enable flag. You can produce the same using that 2 commands.

I understand what you mean, this was my fallback. I thought that OCA has a special docker image for development needs. And I can use it quickly.

**Comment by pedrobaeza** [comment]
No, it uses https://github.com/OCA/maintainer-quality-tools/tree/master/travis. When they were done, there weren't support in Travis for Docker.

**Comment by shepilov-vladislav** [comment]
> No, it uses https://github.com/OCA/maintainer-quality-tools/tree/master/travis. When they were done, there weren't support in Travis for Docker.

@pedrobaeza Please put you attention. It is very interesting why the test fail in past.

Explanation:

in `test_web_login_expire_pass`:

```
two_days_ago = datetime.now() - timedelta(days=2)
...
user.password_write_date = two_days_ago
```

in `_password_has_expired` function:

```
days = (fields.Datetime.now() - self.password_write_date).days
```

and after that days == (`1 day, 23:59:59.824865`).days == 1


I preferred to simply increase the interval to three days in the test. What do you think about that?


Also I added all changes requested by you. So can we merge it now?

**Comment by pedrobaeza** [comment]
OK, if that avoids that edge cases. I remember at least 2 cases in the past with something similar: a test that failed on leap years, and another that fails at midnight when timezone changes the day...

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Thanks for the spent time!

**Comment by shepilov-vladislav** [comment]
> Thanks for the spent time!

Thank you too for help!

**Comment by sbidoul** [comment]
@shepilov-vladislav FYI I fixed the README at e147125d5d82b7b352639b8834d6c884a6c7ca2b. They must be simple text fragments, the rest is generated by the bot which assembles the fragments and add boilerplate text.

**Review by moylop260 — COMMENTED** [review]
:+1:

**Comment by shepilov-vladislav** [comment]
> @shepilov-vladislav FYI I fixed the README at [e147125](https://github.com/OCA/server-auth/commit/e147125d5d82b7b352639b8834d6c884a6c7ca2b). They must be simple text fragments, the rest is generated by the bot which assembles the fragments and add boilerplate text.

👍
